backend/src/main/java/com/bakdata/conquery/models/preproc/Preprocessed.java

```diff
@@ -113,13 +141,39 @@ public void write(HCFile outFile) throws IOException {
 
 		log.debug("Writing Headers");
 
-		writeHeader(outFile.writeHeader());
+		int hash = descriptor.calculateValidityHash(job.getCsvDirectory(), job.getTag());
+
+		PreprocessedHeader header = new PreprocessedHeader(
+				descriptor.getName(),
+				descriptor.getTable(),
+				rows,
+				columns,
+				hash
+		);
+
+		final PreprocessedDictionaries dictionaries = new PreprocessedDictionaries(primaryDictionary, dicts);
+
+		final PreprocessedData data = new PreprocessedData(entityStart, entityLength, columnStores);
+
+
+		try (JsonGenerator generator = createWriter(file)) {
+
+			log.debug("Writing header");
+
+			generator.writeObject(header);
 
-		log.debug("Writing data");
+			log.debug("Writing Dictionaries");
 
-		writeData(outFile.writeContent(), entityStart, entityLength, columnStores, primaryDictionary, dicts);
+			generator.writeObject(dictionaries);
+
+			log.debug("Writing data");
+
+			generator.writeObject(data);
+		}
 	}
```

Comment:
Das sieht sauber aus. Aber das Lesen sollte am besten auch in die selbe Klasse wie das Schreiben

Reply:
Das geht leider nicht weil das lesen noch schritte dazwischen machen muss

Reply:
vielleicht kannst du das in einen autocloseable Reader stecken der `readHeader()`, `readDictionary()` ... und soweiter implementiert und checked das die in der richtigen Reihenfolge aufgerufen werden.